clientLibrary/src/main/scala/za/co/absa/loginclient/tokenRetrieval/service/RetrieveToken.scala

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2023 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.loginclient.tokenRetrieval.service
+
+import com.google.gson.{JsonObject, JsonParser}
+import com.nimbusds.jose.jwk.JWK
+import org.slf4j.{Logger, LoggerFactory}
+import org.springframework.http.{HttpEntity, HttpHeaders, HttpMethod, MediaType, ResponseEntity}
+import org.springframework.web.client.RestTemplate
+import za.co.absa.loginclient.tokenRetrieval.model.{AccessToken, PublicKey, RefreshToken}
+
+import java.net.URLEncoder
+import java.util.Collections
+
+/**
+ * This class is used to retrieve tokens and the public key from the login service.
+ * Public Key comes from a publicly available endpoint. Available in String or JWKS format.
+ * Refresh and Access Keys require authorization. Basic Auth is used for the initial retrieval.
+ * Refresh token from initial retrieval is used to refresh the access token.
+ * @param host The host of the login service.
+ */
+
+case class RetrieveToken(host: String) {
+
+  private val logger: Logger = LoggerFactory.getLogger(this.getClass)
+
+  def getPublicKey: PublicKey = {
+    val issuerUri = s"$host/token/public-key"
+    val jsonString = fetchToken(issuerUri)
+    val token = JsonParser.parseString(jsonString).getAsJsonObject.get("key").getAsString
+    PublicKey(token)
+  }
+
+  def getPublicKeyJwk: JWK = {
+    val issuerUri = s"$host/token/public-key-jwks"
+    val jsonString = fetchToken(issuerUri)
+    val jwkString = JsonParser.parseString(jsonString).getAsJsonObject.get("key").getAsString
+    JWK.parse(jwkString)
+  }
+
+  def fetchAccessToken(username: String, password: String): AccessToken = {
+    fetchAccessAndRefreshToken(username, password)._1
+  }
+
+  def fetchRefreshToken(username: String, password: String): RefreshToken = {
+    fetchAccessAndRefreshToken(username, password)._2
+  }
+
+  def fetchAccessAndRefreshToken(username: String, password: String, groups: Option[List[String]] = None): (AccessToken, RefreshToken) = {
```

Comment:
I would more clearly separate the responsibilities / use-cases of this class, and possibly split it up into multiple classes, or even multiple libraries.

One use-case is for a server application like the Spring Boot applications of AqueDuct or Unify, which protect resources and only need to verify the tokens that the users pass in the authorization header in the REST requests. These are so called resource servers in the oauth2 spec (https://datatracker.ietf.org/doc/html/rfc6749#section-1.1)

Another role in that spec is the "client". In that sense, it is for example the UI of AqueDuct and Unify, but it could also be a lambda function or maybe a Spark application that calls the loginservice to get a token to access the AqueDuct or Unify REST API.

So, `getPublicKey` and `getPublicKeyJwk` would be for resource servers (although it would be covered by NimbusJwtDecoder as mentioned in my other comment), whereas `fetchAccessToken`, `fetchRefreshToken` etc. would be for clients (client in the sense of the OAuth2 spec)

Reply:
Ok, so I split it up into 2 classes. One which is specific to the public key and one that's specific to the tokens. is this what you had in mind?